runtime/compiler/trj9/optimizer/JProfilingValue.cpp

```diff
@@ -204,6 +206,47 @@ TR_JProfilingValue::perform()
    }
 
 void
+TR_JProfilingValue::removeRedundantProfilingValueCalls()
+   {
+   TR::TreeTop *cursor = comp()->getStartTree();
+   TR_BitVector *alreadyProfiledValues = new (comp()->trStackMemory()) TR_BitVector();
+   while (cursor)
+      {
+      TR::Node * node = cursor->getNode();
+      if (node->isProfilingCode() &&
+         node->getOpCodeValue() == TR::treetop &&
+         node->getFirstChild()->getOpCode().isCall() &&
+         (node->getFirstChild()->getSymbolReference()->getReferenceNumber() == TR_jProfile32BitValue ||
+         node->getFirstChild()->getSymbolReference()->getReferenceNumber() == TR_jProfile64BitValue) &&
+         node->getFirstChild()->getNumChildren() == 3)
+         {
+         TR::Node *value = node->getFirstChild()->getFirstChild();
+         if (alreadyProfiledValues->isSet(value->getGlobalIndex()))
+            {
+            /* Found that second and third child were having more than one ref counts. 
+            TR_ASSERT_FATAL(node->getFirstChild()->getSecondChild()->getReferenceCount() == 1 &&
+               node->getFirstChild()->getThirdChild()->getReferenceCount() == 1,
+               "Second and Third Child of the value calls should be referenced only once");
+            */
+            dumpOptDetails(comp(), "%s Removing profiling treetop, node n%dn is already profiled before\n",
+               optDetailString(), value->getGlobalIndex());
+            TR::TransformUtil::removeTree(comp(), cursor);
+            }
+         else
+            {
+            alreadyProfiledValues->set(value->getGlobalIndex());
+            }
+         }
+      // TODO: Following condition will empty down the BitVectors for next Basic Blocks limiting this to  extended basic blocks only
+      // Although it is a safety measure, do we really need to do that?
+      else if (cursor->getNode()->getOpCodeValue() == TR::BBStart && !cursor->getNode()->getBlock()->isExtensionOfPreviousBlock())
```

Comment:
A node cannot appear outside of its extended basic block, but the emptying will keep the number of bits set in the bitvector low so perhaps that is a justification. It may also make debugging easier.